### ISSUE TITLE
Flow unconditionally does raster-cache related work for `CacheableContainerLayer`, which is wasted if the raster cache is disabled.

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#133377]
See https://github.com/flutter/engine/pull/45131#issuecomment-1694089281, shows up in profiles.

We should make sure this gets guarded on whether the raster cache is enabled or not.

@flar @JsouLiang @jonahwilliams

**Comment by flar** [comment]
Wait, LRCI::Make is showing up? With 1ms+ per frame?

That's way out of whack for just constructing an object with a couple of fields. Something looks wrong there.

Or is this a total cost over a large number of frames?

**Comment by dnfield** [comment]
It's not taking 1ms per frame, it's 1ms aggregated over many frames.

It's doing allocations that aren't necessary.

**Comment by flar** [comment]
We could create a dummy singleton "cache is disabled raster cache item" and simply return that from the Make method when the cache is disabled. It will have to be changed from "unique_ptr" to allow sharing.

**Comment by jonahwilliams** [comment]
Is it necessary to set this item up in the constructor? Could this be lazily computed and initialized if the raster cache is enabled? I'm not sure that we should spend much time thinking about this, but it is evidence that we probably need to rethink flow for an Impeller only world.

**Comment by flar** [comment]
Hmmm, these are created once when a layer is created and then reused so it can't be permanently disabled in Make in case this is a retained layer. That makes it messy.

**Comment by flar** [comment]
Or perhaps cacheableContainer could just *be* the LayerRCI. I'd have to look at the code more deeply, but it's not clear why we need to allocate an object to do this work when the object just stores a backpointer to the layer and a couple of config constants.

**Comment by jonahwilliams** [comment]
(presumably there is a lot more we could cut out)

**Comment by flar** [comment]
I think we are prematurely micro-optimizing in a lot of cases. Viewing the abstraction layers as enemies rather than just maintaining them to be efficient starts to become its own tech debt.

**Comment by dnfield** [comment]
@flar can we move the `LayerRasterCacheItem::Make` call to the `raster_cache_item` getter like this?

```cpp
  const LayerRasterCacheItem* raster_cache_item() const {
    if (!layer_raster_cache_item_) {
      layer_raster_cache_item_ = LayerRasterCacheItem::Make(this, layer_cached_threshold_, can_cache_children_);
    }
    return layer_raster_cache_item_.get();
  }
```

**Comment by dnfield** [comment]
Ah, it looks like that is only really used in tests. 

We could alternatively just add the bool on the ctor of this class, so that you have to construct it knowing whether the cahce is enabled or not to avoid creating the item...

**Comment by dnfield** [comment]
> I think we are prematurely micro-optimizing in a lot of cases. Viewing the abstraction layers as enemies rather than just maintaining them to be efficient starts to become its own tech debt.

If there's no raster caching, what benefit does Flow give us as an abstraction?

I think you're more up to date on this than I am, but we've been advised before by some Android folks to drop composition layers where we can. My argument at the time was that we benefit from raster caching and Flow enables that significantly. But if the raster cache is off, what are the reasons to keep flow?

**Comment by jonahwilliams** [comment]
We could probably have a much simpler flow, it allows us to cache and reuse impeller pictures. We don't benefit from that at all due to the complete redispatch from display list to impeller on each frame, but we could

**Comment by flar** [comment]
> > I think we are prematurely micro-optimizing in a lot of cases. Viewing the abstraction layers as enemies rather than just maintaining them to be efficient starts to become its own tech debt.
> 
> If there's no raster caching, what benefit does Flow give us as an abstraction?

Why not just fork the entire engine? Have an impeller-specific engine and a legacy engine?

You are making a major structural change to an existing engine that works fine based on a very tiny perf issue that can be fixed. That's what worries me.

**Comment by flar** [comment]
> We could probably have a much simpler flow, it allows us to cache and reuse impeller pictures. We don't benefit from that at all due to the complete redispatch from display list to impeller on each frame, but we could

We can have a much simpler everything when Impeller can do everything the current engine does with Skia. We aren't there yet. Rather than fork our development over tiny things that can be fixed, let's look at ways to minimize the impact in the near term.

**Comment by jonahwilliams** [comment]
No one is proposing forking the engine because of a std::unique_ptr allocation.

**Comment by dnfield** [comment]
So let's table the flow-related discussion for this issue. I think it'd be beneficial to sort out what purpose flow serves and why we want to keep it or not as an abstraction.

For this issue, I think the right answer is going to be adding a constructor parameter to `CacheableContainerLayer` that feeds in whether the raster cache is actually enabled or not, so it can avoid the extra allocations when it is disabled. WDYT of that idea @flar?

**Comment by flar** [comment]
I think lazy is even better. Just have the layers call an accessor to get to the RCI and the lazy (non-)construction can be hidden behind it. It could even return a pointer to a singleton "quickly say no to everything" implementation of RCI for minimal code impact.

**Comment by flar** [comment]
(Or the call sites could just take "null" for an answer and skip the work)

**Comment by flar** [comment]
I did a quick test.

I measured opacity_peephole_grid_of_opacity_perf__e2e_summary and then made a change to CacaheableLayer. In the constructor where it constructs the LayerRCI I put it in a loop that did it 1000x. The layers per frame were about 80 and then over 600 frames we were allocating around 50k of these objects. With the loop we were allocating almost 50m of them.

The average frame build time went from ~6ms to just over 10ms indicating that 1000x the load of creating these objects costs about 4ms over all of the frames - which means the cost is around 4us per frame. That's quite a bit more than I expected, but 4us*80 = 320us per frame for this benchmark is starting to seem like something to look into, but not a huge performance drain in the absolute. I'll see if I can quickly make them lazy to non-existant...
